test/jdk/javax/swing/JTableHeader/TableHeaderBorderPositionTest.java

```diff
@@ -65,64 +64,40 @@ private static void test() {
-        try {
-            header.paint(g2d);
-        } finally {
-            g2d.dispose();
-        }
 
-        w = (int)Math.ceil(SCALE * WIDTH);
-        h = (int)Math.ceil(SCALE * HEIGHT);
-        BufferedImage imgData = new BufferedImage(w, h,
-                BufferedImage.TYPE_INT_RGB);
-        g2d = imgData.createGraphics();
-        g2d.scale(SCALE, SCALE);
         try {
+            header.paint(g2d);
+            g2d.translate(0, header.getHeight());
             table.paint(g2d);
         } finally {
             g2d.dispose();
         }
 
-        int verticalLineCol = (int)(table.getTableHeader().
-                getColumnModel().getColumn(0).getWidth() * SCALE) - 2;
-        int expectedRGB = imgData.getRGB(verticalLineCol, 0);
+        int verticalLineCol = (int) (table.getTableHeader()
+                                          .getColumnModel()
+                                          .getColumn(0)
+                                          .getWidth() * SCALE) - 2;
+        int expectedRGB = bufferedImage.getRGB(verticalLineCol, 0);
+        int maxHeight = (int)(((double)table.getTableHeader().getHeight() * SCALE )
+                + ((double)table.getRowHeight() * SCALE));
 
-        for (int y = 0; y < imgHeader.getHeight(); y++) {
+        for (int y = 0; y < maxHeight; y++) {
             for (int x = verticalLineCol; x < verticalLineCol + 3; x++) {
-                if (expectedRGB != imgHeader.getRGB(x, y)) {
-                    saveImage(imgHeader, "FailureImageHeader.png");
-                    throw new RuntimeException("Header Comparison Test " +
-                            "Failed at <" + x + ", " + y + ">");
-                }
-            }
-        }
-
-        int maxVerticalLine = (int) Math.ceil(
-                table.getRowCount() * table.getRowHeight() * SCALE);
-        for (int y = 0; y < maxVerticalLine; y++) {
-            for (int x = verticalLineCol; x < verticalLineCol + 3; x++) {
-                if (expectedRGB != imgData.getRGB(x, y)) {
-                    saveImage(imgData,"FailureImageData.png");
-                    throw new RuntimeException("TableData Comparison Test " +
-                            "Failed at <" + x + ", " + y + ">");
+                if (expectedRGB != bufferedImage.getRGB(x, y)) {
+                    try {
+                        ImageIO.write(bufferedImage, "png",
+                                new File("FailureImage.png"));
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                    throw new RuntimeException("Test Failed at <" + x + ", " + y + ">");
                 }
             }
         }
         System.out.println("Test Passed");
     }
-
-    private static void saveImage(BufferedImage img, String fileName) {
-        try {
-            ImageIO.write(img, "png", new File(fileName));
-        } catch (IOException e) {
-            e.printStackTrace();
-        }
-    }
```

Comment:
In fact, it looks better with `saveImage` method.

Reply:
Since we are saving only once, I don't think separate method is required.

Reply:
It is not required but it separates the duties: the `test` method *verifies* the rendering, it calls `saveImage` to save the image. Such a design is in the spirit of OOP.

As bonus, it allows one to save the image easily before it is verified or even after each paint is performed. Thus, it has its benefits.

I'm not insisting, even though I think it's a better design.

Reply:
Ok, no issues, If it adds something better then I'll update it.

Reply:
> Ok, no issues, If it adds something better then I'll update it.

The test in #14464, `JTableHeaderLabelRightAlignTest.java` will also benefit from it, and you haven't integrated it yet.